[Articles] Getting started setting up Neovim for Swift (#675)
apple/swift-org-website: documentation/articles/zero-to-swift-nvim.md

Reviewer: @etcwilde - Paris requested I TAL at this and provide some very high-level documentation writing comments. Plmk if you have questions. :-)

1. Installing Neovim.
2. Installing `lazy.nvim` to manage our plugins.
3. Configuring the SourceKit-LSP server.
4. Setting up Language-Server-driven autocompletion with _nvim-cmp_.
5. Setting up snippets with _LuaSnip_.

The following sections are provided to help guide you through the setup:

- [Prerequisites](#prerequisites)
- [Package Management](#packaging-with-lazy)
- [Language Server Support](#language-server-support)
    - [File Updates](#file-updating)
- [Autocomplete](#auto-complete)
- [Snippets](#Snippets)

> Tip: If you already have Neovim, Swift, and a package manager installed, you can skip to setting up [Language Server support](#language-server-support).

> Note: If you are bypassing the [Prerequisites](#prerequisites) section, make sure your
copy of Neovim is version v0.9.4 or higher, or you may experience issues with some
of the Language Server Protocol (LSP) Lua APIs.

## Prerequisites

To get started, you'll need to install Neovim. The Lua
APIs exposed by Neovim are under rapid development. We will want to take
advantage of the recent improvements in the integrated support for Language
Server Protocol (LSP), so we will need a fairly recent version

> of Neovim. I'm running Ubuntu 22.04 on an `x86_64` machine. Unfortunately, the

Reviewer: Recommend starting a new paragraph with: "I'm running Ubuntu 22.24..."

version of Neovim shipped in the Ubuntu 22.04 `apt` repository is too old to

> support many of the APIs that we will be using. For this install, I used `snap`

Reviewer: Also Recommend starting a new paragraph with: "For this install, I used `snap`..."

```

## Getting Started

We have working copies of Neovim and Swift on our path. While we can start with
a `vimrc` file, Neovim is transitioning from using vimscript to Lua. Lua
is easier to find documentation for since it's an actual programming language,
tends to run faster, and pulls your configuration out of the main runloop so
your editor stays nice and snappy.
You can still use a `vimrc` with vimscript, but we'll use Lua.

The main Neovim configuration file goes in `~/.config/nvim`. The other Lua files
go in `~/.config/nvim/lua`. Go ahead and create an `init.lua` now;

```console
 $  mkdir -p ~/.config/nvim/lua && cd ~/.config/nvim
 $  nvim init.lua
```

> Note: The examples below contain a GitHub link to the plugin to help you readily access the documentation. You can also explore the plugin itself.

## Packaging with _lazy.nvim_

While it's possible to set everything up manually, using a package manager helps
keep your packages up-to-date, and ensures that everything is installed
correctly when copy your configuration to a new computer. Neovim also has a
built-in plugin manager, but I have found
[_lazy.nvim_](https://github.com/folke/lazy.nvim) to work well. We will

> start with a little bootstrapping script to install _lazy.nvim_ if it isn't

Reviewer: Recommend starting a new paragraph with: "We will start with a little bootstrapping...."

        "https://github.com/folke/lazy.nvim.git",
        "--branch=stable",
        lazypath
    })
end
vim.opt.rtp:prepend(lazypath)
```

This snippet clones _lazy.nvim_ if it doesn't already exist, and then adds it to the
runtime path. Now we initialize _lazy.nvim_ and tell it where to look for the plugin
specs.

```lua
require("lazy").setup("plugins")
```

This configures _lazy.nvim_ to look in a `plugins/` directory under our `lua/`
directory for each plugin. We'll also want a place to put our own non-plugin
related configurations, so we'll stick it in `config/`. Go ahead and create
those directories now.

```console
 $  mkdir lua/plugins lua/config
```

See [lazy.nvim Configuration](https://github.com/folke/lazy.nvim?tab=readme-ov-file#%EF%B8%8F-configuration) for details on configuring _lazy.nvim_.

![_lazy.nvim_ package manger](/assets/images/zero-to-swift-nvim/Lazy.png)

Note that your configuration won't look exactly like this.
We have only installed _lazy.nvim_, so that is the only plugin that is listed on
your configuration at the moment.
That's not very exciting to look at, so I've added a few additional plugins to
make it look more appealing.

> To check that it's working, launch Neovim. You should first see an error saying

Reviewer: The last nit ... consider formatting for easier readability:

"To check that it's working:
1. Launch Neovim. 
You should first see an error saying that there were no specs found for module plugins. This just means there aren't any plugins. 
2. Press Enter and type, `:Lazy`. 
_lazy.nvim_ will list the plugins loaded. There should only be one right now: "lazy.nvim". This is _lazy.nvim_ tracking and updating itself. 
3. From the _lazy.nvim_ menu, press `I` to install new plugins. 
`U` updates the plugins and `X` deletes any plugins that _lazy.nvim_ installed, but are no longer being tracked in your configuration."

Author: There, I broke up apart the last point a bit further so that each button is a separate element.
Something I'm noticing, if I try to number them, the website generator seems very intent on renumbering all of the points as `1.` so I left them as an unordered list. Will that work?

To check that it's working:
 - Launch Neovim.

   You should first see an error saying that there were no specs found for
   module plugins. This just means that there aren't any plugins.

 - Press Enter and type, `:Lazy`.

   _lazy.nvim_ lists the plugins installed. There should only be one right now:
   "lazy.nvim". This is _lazy.nvim_ tracking and updating itself.

 - We can manage out plugins through the _lazy.nvim_ menu.
    - Pressing `I` will install new plugins.
    - Pressing `U` will update installed plugins.
    - Pressing `X` will delete any plugins that _lazy.nvim_ installed, but are
      no longer tracked in your configuration.

Reviewer: That works. Thanks.

```

This snippet clones `lazy` if it doesn't already exist, and then adds it to the
runtime path. Now we initialize lazy and tell it where to look for the plugin
specs.

```lua
require("lazy").setup("plugins")
```

This configures `lazy` to look in a `plugins/` directory under our `lua/`
directory for each plugin. We'll also want a place to put our own non-plugin
related configurations, so we'll stick it in `config/`. Go ahead and create
those directories now.

```sh
 $  mkdir lua/plugins lua/config
```

See [lazy.nvim Configuration](https://github.com/folke/lazy.nvim?tab=readme-ov-file#%EF%B8%8F-configuration) for details on configuring `Lazy`.

![Lazy package manger](/assets/images/zero-to-swift-nvim/Lazy.png)

Note that your configuration won't look exactly like this.
Since we have only installed `lazy.nvim`, that's all that `Lazy` will list.
That's not very exciting to look at, so I've added a few additional plugins to
make it look more appealing.

To check that it's working, launch Neovim. You should first see an error saying
that there were no specs found for module plugins. This just means that we don't
have any plugins. Press <ENTER>, and then type `:Lazy`.

> `Lazy` will list the plugins loaded.

Reviewer: I believe "Lazy" is the product name here. Recommend removing the backticks and just writing it as Lazy. However, if/when referring to it in UI terms, put it in backticks.

There should only be one right now, "lazy.nvim".
This is Lazy tracking and updating itself.

> From the `Lazy` menu, pressing `I` will install new plugins, `U` will update the

Reviewer: I believe Lazy here is used as a product name in this sentence. Recommend writing as Lazy and not 'Lazy'.

version of Neovim shipped in the Ubuntu 22.04 `apt` repository is too old to
support many of the APIs that we will be using. For this install, I used `snap`
to install Neovim v0.9.4.

```sh
 $  sudo snap install nvim --classic
 $  nvim --version
NVIM v0.9.4
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
Compilation: /usr/bin/cc -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-pa...

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

You may need to adjust the commands you use based on what distribution and
operating system you are using.
Alternatively, the Neovim GitHub repository has downloadable packages with the latest versions.
See
[install from download](https://github.com/neovim/neovim/blob/master/INSTALL.md#install-from-download)
for more information.

Once we have Neovim installed, we'll install a Swift toolchain. Go over to the
download page on [swift.org](download.swift.org) and download the release Swift
tarball. This package contains the runtime libraries, compilers, debugger, and
integrated development environment (IDE) tools needed for a Swift development
experience.

> ```sh

Reviewer 2: I'm not sure this is on the list of supported languages for the syntax highlighter, it should probably be
```suggestion
```bash
```

---
layout: page
date: 2024-06-04 15:13:07
title: Getting Started with Swift in Neovim
author: [etcwilde]
---

Neovim is a modern reimplementation of the *vi text editor*, with a lot of new functionality piled
on top including asynchronous operations and Lua bindings for a snappy editing
experience. We will look at configuring a new install of Neovim for Swift

> development on Ubuntu 22.04. I have pre-installed the Solarized theme so the

Reviewer 2: I'm assuming this works on other flavours of Linux (and Windows and macOS as well? Would be good to clarify that here

We will start by installing Neovim and Swift and then install `lazy.nvim` to
manage our plugins. Then we'll configure the SourceKit-LSP server, setup
Language-Server-driven autocompletion with `nvim-cmp`, and snippets with
`LuaSnip`.

- [Prerequisites](#prerequisites)
- [Package Management](#packaging-with-lazy)
- [Language Server Support](#language-server-support)
    - [File Updates](#file-updating)
- [Autocomplete](#auto-complete)
- [Snippets](#Snippets)

If you already have Neovim, Swift, and a package manager installed, you can skip
down to setting up [Language Server support](#language-server-support).

## Prerequisites

To get started, you'll need to install Neovim and a Swift toolchain. The Lua
APIs exposed by Neovim are under rapid development. We will want to take
advantage of the recent improvements in the integrated support for Language
Server Protocol (LSP), so we will need a fairly recent version
of Neovim. I'm running Ubuntu 22.04 on an `x86_64` machine. Unfortunately, the
version of Neovim shipped in the Ubuntu 22.04 `apt` repository is too old to
support many of the APIs that we will be using. For this install, I used `snap`
to install Neovim v0.9.4.

```bash

>  $  sudo snap install nvim --classic

Reviewer 2: It might be nice to also mention other installation methods for different platforms?

Author: I moved the link to the Neovim install page up above this block so that folks are aware of it. See what you think.

 $  nvim --version
NVIM v0.9.4
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
Compilation: /usr/bin/cc -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-pa...

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

You may need to adjust the commands you use based on what distribution and
operating system you are using.
Alternatively, the Neovim GitHub repository has downloadable packages with the latest versions.
See
[install from download](https://github.com/neovim/neovim/blob/master/INSTALL.md#install-from-download)
for more information.

Once we have Neovim installed, we'll install a Swift toolchain. Go over to the
download page on [swift.org](download.swift.org) and download the release Swift
tarball. This package contains the runtime libraries, compilers, debugger, and
integrated development environment (IDE) tools needed for a Swift development
experience.

```bash
 $  cd ~/Downloads
 $  wget https://download.swift.org/swift-5.10-release/ubuntu2204/swift-5.10-RELEASE/swift-5.10-RELEASE-ubuntu22.04.tar.gz

>  $  tar xf swift-5.10-RELEASE-ubuntu22.04.tar.gz

Reviewer 2: Not a massive fan of the leading `$` as it makes it harder to copy commands

Author: Hmm, I was using the prompt to indicate what was command vs output from the command (at least in the blocks above and below). Should I not include the output and just have the commands then? 

Some sites know how to drop the leading prompt character so copy-paste only copies the commands themselves and not the prompt character or the output from the command, is there a way to indicate to swift.org what should by copied and shouldn't be in a code block?
Another school of thought is that copy-pasting commands is a security anti-pattern since it has been used for malicious attacks in the past.

Reviewer 2: Yeah we don't have a good story for that at the moment unfortunately

Reviewer 3: We can just turn off selection for the prompt: https://github.com/apple/swift-org-website/pull/694

This will work if you use

    ```console
    $ some-command -lots -of -options
    Output goes here
    ```

which will render as

```console
$ some-command -lots -of -options
Output goes here
```

but the `$ ` at the beginning will no longer be selectable.

```

> That drops a `swift-5.10-RELEASE-ubuntu22.04/usr` directory in our `~/Downloads`

Reviewer 2: We should link to the Install Swift page on Swift.org as that contains a comprehensive guide for all the different platforms, especially once #676 is landed.

Reviewer 4: Yeah, IMO this entire section can essentially just become a link to "install Swift"

but should be fine for now. We'll need to add the compiler and tools to our
path.

```bash
 $  export PATH="$PWD/swift-5.10-RELEASE-ubuntu22.04/usr/bin:$PATH"
 $  which swiftc
/home/ewilde/Downloads/swift-5.10-RELEASE-ubuntu22.04/usr/bin/swiftc
 $  swiftc --version
Swift version 5.10 (swift-5.10-RELEASE)
Target: x86_64-unknown-linux-gnu
```

## Getting Started

We have working copies of Neovim and Swift on our path. While we can start with
a `vimrc` file, Neovim is transitioning from using vimscript to Lua. Lua
is easier to find documentation for since it's an actual programming language,
tends to run faster, and pulls your configuration out of the main runloop so
your editor stays nice and snappy.
You can still use a `vimrc` with vimscript, but we'll use Lua.

The main Neovim configuration file goes in `~/.config/nvim`. The other Lua files
go in `~/.config/nvim/lua`. Go ahead and create an `init.lua` now;

```bash
 $  mkdir -p ~/.config/nvim/lua && cd ~/.config/nvim
 $  nvim init.lua
```

> Note: The examples below contain a GitHub link to the plugin to help you readily access the documentation. You can also explore the plugin itself.

> ## Packaging with _lazy.nvim_

Reviewer 2: Is this deliberately italicised? Feels like monospaced would be better

Author: Yeah, that's the name of the plugin.

            }),
            snippets = {
                expand = function(args)
                    luasnip.lsp_expand(args)
                end,
            },
        }
        cmp.setup(opts)
    end,
},
```

Now our tab-key is thoroughly overloaded in super-tab fashion.
 - If the completion window is open, pressing tab selects the next item in the
   list.
 - If you press tab over a snippet, the snippet will expand, and continuing to
   press tab moves the cursor to the next selection point.
 - If you're neither auto-completing nor expanding a snippet, it will behave
   like a normal `tab` key.

Now we need to write up some snippets. LuaSnip supports several snippet formats,
including a subset of the popular
[TextMate](https://macromates.com/textmate/manual/snippets),
[Visual Studio Code](https://code.visualstudio.com/docs/editor/userdefinedsnippets) snippet format,
its own [Lua-based](https://github.com/L3MON4D3/LuaSnip/blob/master/Examples/snippets.lua) API,
and snippets coming from an [LSP server](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#snippet_syntax).

Here are some snippets that I've found to be useful:

> ```snipmate

Reviewer 2: As an FYI there's no syntax highlighting for this file, but I can't see it as a supported library in Highlight.js

Author: That's fine I suppose. Not much I can do about that. At least it'll highlight correctly if someone looks at the markdown file in the repo. 😅

```

This configures _lazy.nvim_ to look in a `plugins/` directory under our `lua/`
directory for each plugin. We'll also want a place to put our own non-plugin
related configurations, so we'll stick it in `config/`. Go ahead and create
those directories now.

```console
 $  mkdir lua/plugins lua/config
```

See [lazy.nvim Configuration](https://github.com/folke/lazy.nvim?tab=readme-ov-file#%EF%B8%8F-configuration) for details on configuring _lazy.nvim_.

![_lazy.nvim_ package manger](/assets/images/zero-to-swift-nvim/Lazy.png)

Note that your configuration won't look exactly like this.
We have only installed _lazy.nvim_, so that is the only plugin that is listed on
your configuration at the moment.
That's not very exciting to look at, so I've added a few additional plugins to
make it look more appealing.

To check that it's working:
 - Launch Neovim.

   You should first see an error saying that there were no specs found for
   module plugins. This just means that there aren't any plugins.

 - Press Enter and type, `:Lazy`.

   _lazy.nvim_ lists the plugins installed. There should only be one right now:
   "lazy.nvim". This is _lazy.nvim_ tracking and updating itself.

>  - We can manage out plugins through the _lazy.nvim_ menu.

Reviewer: Oops! I think "out" should be "our"?

Author: Yep!

advantage of the recent improvements in the integrated support for Language
Server Protocol (LSP), so we will need a fairly recent version
of Neovim. I'm running Ubuntu 22.04 on an `x86_64` machine. Unfortunately, the
version of Neovim shipped in the Ubuntu 22.04 `apt` repository is too old to
support many of the APIs that we will be using. For this install, I used `snap`
to install Neovim v0.9.4.

```bash
 $  sudo snap install nvim --classic
 $  nvim --version
NVIM v0.9.4
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
Compilation: /usr/bin/cc -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-pa...

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

You may need to adjust the commands you use based on what distribution and
operating system you are using.
Alternatively, the Neovim GitHub repository has downloadable packages with the latest versions.
See
[install from download](https://github.com/neovim/neovim/blob/master/INSTALL.md#install-from-download)
for more information.

Once we have Neovim installed, we'll install a Swift toolchain. Go over to the

> download page on [swift.org](https://download.swift.org) and download the release Swift

Reviewer 2: The download page has now been deprecated and no longer linked to on the homepage in favour of the new download page

---
layout: page
date: 2024-06-04 15:13:07
title: From Zero to Swift | Configuring Neovim for Swift Development
author: [etcwilde]
---

> Neovim is a modern reimplementation of _Vi IMproved (vim)_, a popular terminal-based text

Author: I have mixed feelings on expanding vim. It's technically not wrong, but I don't think it helps with comprehension and makes it a little awkward to read. Thoughts? @al45tair, what do you think? Maybe I've been looking at this for too long?

Reviewer: Ack. The recommendation is based on standard writing style guidelines. Exceptions can be made if your target audience already understands the terminology and definition.

Reviewer 3: I'd probably make sure that I spelled it Vim, rather than in all lowercase, unless it's a command in which case `vim`. I don't think I'd expand it — I think it's a name at this point. It's also a little tricky to expand, because `vi` was a shortening of **VI**sual, so `vim` is a play on words — it's both **VI**sual and **IM**proved.

In any event, I think the audience is likely familiar with Vim.

Reviewer 3: (Having read through it, a few times, I don't _mind_ the expansion either, so if you want to leave it how it is, that's fine by me.)

Author: _Neo-VIsual-IMproved (Neovim)_ intensifies. No, I think that looks silly and I think that will actually confuse more people than it helps.

I'll go ahead and capitalize the name though.

Reviewer: I'm good with the majority vote. Thanks!

# From Zero to Swift | Getting started with Swift in Neovim

Neovim is a modern reimplementation of vi, with a lot of new functionality piled
on top including asynchronous operations and lua bindings for a snappy editing
experience. We'll look at configuring a new install of Neovim for Swift
development on Ubuntu 22.04. I have preinstalled the Solarized theme so the
colors may not match yours exactly.

We will start by installing Neovim and Swift and then install `Lazy.nvim` to
manage our plugins. Then we'll configure the SourceKit LSP server, setup
LSP-driven autocompletion with `nvim-cmp`, and snippets with `LuaSnip`. Then
we'll finish with some autocommands that are useful if you need to read

> swiftinterface files or find yourself reading SIL. If you already have Neovim,

Reviewer: is swiftinterface a UI term? If so, consider putting it in backticks. Spell out all acronyms the first time it's mentioned for clarity. For example:
- SourceKit Language Server Protocol (LSP) server
- SIL should be written as: Swift Integration Layer (SIL)

Then you can use just the acronym throughout the documentation.

Author: Ended up removing the autocmd bit. `SIL` is really something that only a compiler engineer working on Swift will care about. If folks want it back, I can bring it back, but I don't know that it's worth having just for swift interfaces.

Swift, and a package manager installed, you can skip down to setting up
[Language Server support](#language-server-support).

> 

Reviewer: Add a lead-in sentence for this list. Maybe something like, "Here's a list of topics that we'll cover in this guide:"

Author: That's what I was trying to do with this

> We will start by installing Neovim and Swift and then install `lazy.nvim` to
 manage our plugins. Then we'll configure the SourceKit-LSP server, setup
 Language-Server-driven autocompletion with `nvim-cmp`, and snippets with
 `LuaSnip`.

Reviewer: Noted.

- [Prerequisites](#prerequisites)
- [Package Management](#packaging-with-lazy)
- [Language Server](#language-server-support)
    - [File Updates](#file-updating)
- [Autocomplete](#auto-complete)
- [Snippets](#Snippets)
- [Autocommands](#auto-commands)

## Prerequisites

We will want the Lua APIs exposed in fairly recent versions of Neovim, and we'll
need a Swift toolchain. I'm running Ubuntu 22.04 on an X86 machine, so the
commands you use may need to vary depending on your OS. Recent versions of
Neovim have better LSP support, but this means that we need a newer version than
what is in the Ubuntu 22.04 repository. For this install, I used `snap` to
install Neovim v0.9.4.

```sh
 $  sudo snap install nvim --classic
 $  nvim --version
NVIM v0.9.4
Build type: RelWithDebInfo
LuaJIT 2.1.1692716794
Compilation: /usr/bin/cc -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-pa...

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

Alternatively, the neovim github repository has downloadable packages with the
latest versions. Instructions are available at

> https://github.com/neovim/neovim/blob/master/INSTALL.md#install-from-download

Reviewer: Consider embedding long GitHub links using the specific name of the section to add clarity. For example in the sentence, "Alternatively, the neovim GitHub repository has downloadable packages with the
latest versions. See [Install from download](https://github.com/neovim/neovim/blob/master/INSTALL.md#install-from-download) for more information."

Once we have neovim installed, we'll install a Swift toolchain. Go over to
the download page on [swift.org](download.swift.org) and download the release
Swift tarball. This package contains the runtime libraries, compilers, debugger,

> and IDE tools needed for a Swift development experience.

Reviewer: Spell out Integrated Development Environment (IDE) on first mention.

```sh
 $  cd ~/Downloads
 $  wget https://download.swift.org/swift-5.10-release/ubuntu2204/swift-5.10-RELEASE/swift-5.10-RELEASE-ubuntu22.04.tar.gz
 $  tar xf swift-5.10-RELEASE-ubuntu22.04.tar.gz
```

> That drops a `swift-5.10-RELEASE-ubuntu22.04/usr` directory in our Downloads

Reviewer: Is Downloads part of the UI? If so, consider changing to `Downloads`

directory. The downloads folder isn't a great place to keep things permanently,
but should be fine for now. We'll need to add the compiler and tools to our
path.

```sh
 $  export PATH="$PWD/swift-5.10-RELEASE-ubuntu22.04/usr/bin:$PATH"
 $  which swiftc
/home/ewilde/Downloads/swift-5.10-RELEASE-ubuntu22.04/usr/bin/swiftc
 $  swiftc --version
Swift version 5.10 (swift-5.10-RELEASE)
Target: x86_64-unknown-linux-gnu
```

## Getting Started

> We have working copies of neovim and Swift on our path. While we can start with

Reviewer: Capital **N**eovim versus lower-case **n**eovim?

a `vimrc` file, Neovim is pushing away from the vimscript over to using lua. Lua
is easier to find documentation for since it's an actual language, tends to run
faster, and pulls your configuration out of the main runloop so your editor
stays nice and snappy. You can still use a `vimrc` with vimscript, but we'll use
lua.

The main lua Neovim configuration file goes in `~/.config/nvim`. The other lua
files go in `~/.config/nvim/lua`. Go ahead and create an `init.lua` now;
```sh
 $  mkdir -p ~/.config/nvim/lua && cd ~/.config/nvim
 $  nvim init.lua
```

## Packaging with Lazy

https://github.com/folke/lazy.nvim

> While it's possible to set everything up by hand, using a package manager helps

Reviewer: Consider changing "While it's possible to set everything up by hand," to "While it's possible to set everything up manually,"

keep your packages up-to-date, and ensures that everything is installed

> correctly when copy your configuration to a new computer. Neovim has a builtin

Reviewer: Think this is built-in with a hyphen

> plugin management support, but I have found `lazy.nvim` to work well, so we'll

Reviewer: Same goes for plug-in

> start with a little bootstrapping script to install lazy if it isn't already,

Reviewer: is **lazy** a UI term? If so, consider writing it as `lazy`

Reviewer: Ack. You may want to italicize it to call it out from the rest of the text.

add it to our runtime path, and finally configure our packages.

At the top of your `init.lua` write:
```lua
local lazypath = vim.fn.stdpath("data") .. "/lazy/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
    vim.fn.system({
        "git",
        "clone",
        "--filter=blob:none",
        "https://github.com/folke/lazy.nvim.git",
        "--branch=stable",
        lazypath
    })
end
vim.opt.rtp:prepend(lazypath)
```

This snippet clones `lazy` if it doesn't already exist, and then adds it to the
runtime path. Now we initialize lazy and tell it where to look for the plugin
specs.

```lua
require("lazy").setup("plugins")
```

This configures lazy to look in a `plugins/` directory under our `lua/`
directory for each plugin. We'll also want a place to put our own non-plugin
related configurations, so we'll stick it in `config/`. Go ahead and create
those directories now.

```sh
 $  mkdir lua/plugins lua/config
```

You can read more about setting up Lazy on their github page;

> https://github.com/folke/lazy.nvim?tab=readme-ov-file#%EF%B8%8F-configuration.

Reviewer: Consider embedding this link using the name of the actual content it renders: [lazy.nvim Configuration](https://github.com/folke/lazy.nvim?tab=readme-ov-file#%EF%B8%8F-configuration)

![Lazy package manger](/assets/images/zero-to-swift-nvim/Lazy.png)

Note that your configuration won't look exactly like this. It will likely only
list `lazy.nvim` since that is all that is installed right now, which isn't very
exciting to look at, so I've added a few plugins to make it look more appealing.

> To check that it's working, launch neovim. You should first see an error saying

Reviewer: Neovim (upper case?)

> that there were no specs found for module plugins. This just means that we don't

Reviewer: plug-ins instead of plugin?

Author: It's pretty common to omit the hyphen in `plugin`. Neovim's own documentation omits the hyphen in `plugin` so I'd prefer to match that.

Reviewer: Noted to align with Neovim's documentation.

have any plugins. Press <ENTER>, and then type `:Lazy`. Lazy will list the

> plugins loaded. There should only be one right now, "lazy.nvim". This is Lazy

Reviewer: is "lazy.nvim" a UI term here? If so, can write as: `lazy.nvim`

Author: "lazy.nvim" is the full name of the plugin itself. I've changed all of the instances of "Lazy" to _lazy.nvim_ to match the name in the plugin documentation and readme page.

Go ahead and create a new file under `lua/plugins/lsp.lua`. In it, we'll start
by adding the following snippet.

```lua
return {
    {
        "neovim/nvim-lspconfig"
        config = function()
            local lspconfig = require('lspconfig')
            lspconfig.sourcekit.setup {}
        end,
    }
}
```

While this gives us LSP support through sourcekit-lsp, there are no keybindings,
so it's not very practical. Lets hook those up now.

We'll set up an auto command that fires when LSP attaches in the `config`
function under where we set up the `sourcekit` server. The keybindings are
applied to all LSP servers so you end up with a consistent experience across
languages.

```lua
config = function()
    local lspconfig = require('lspconfig')
    lspconfig.sourcekit.setup {}

    vim.api.nvim_create_autocmd('LspAttach', {
        desc = 'LSP Actions',
        callback = function(args)
            vim.keymap.set('n', 'K', vim.lsp.buf.hover, {noremap = true, silent = true})
            vim.keymap.set('n', 'gd', vim.lsp.buf.definition, {noremap = true, silent = true})
        end,
    })
end,
```

![LSP powered live error messages](/assets/images/zero-to-swift-nvim/LSP-Error.png)

> I've created a little example Swift package that computes Fibonacci numbers

Reviewer: Maybe consider linking to an official/credible resource that explains Fibonacci numbers for users who aren't familiar with it?

asynchronously. Pressing `shift` + `k` on one of the references to the
`fibonacci` function shows the documentation for that function, along with the
function signature. The LSP integration is also showing that we have an error in
the code.

### File Updating

> SourceKit-LSP is increasingly relying on the editor informing the server when

Reviewer 3: ```suggestion
SourceKit-LSP increasingly relies on the editor informing the server when
```

> certain files change. This need is communicated through "dynamic registration".

Reviewer 3: IMO italics look nicer than quotes for definitions.

```suggestion
certain files change. This need is communicated through _dynamic registration_.
```

We will use `nvim-cmp` to act as the autocomplete mechanism. We'll start by
telling Lazy to download the package and to load it lazily when we enter insert
mode since you don't need autocompletion if you're not editing the file.

```lua
-- lua/plugins/autocomplete.lua
return {
    {
        "hrsh7th/nvim-cmp",
        version = false,
        event = "InsertEnter",
    },
}
```

`nvim-cmp` doesn't come with completion sources to start with, those are
additional plugins. We'll add the LSP source as a dependency to provide
LSP-powered completions, and tell Lazy that the completions depend on it so that
those also get loaded when we enable auto-completions. I'm also a fan of having
path expansions and expansions from text in the same buffer.

```lua
-- lua/plugins/autocomplete.lua
return {
    {
        "hrsh7th/nvim-cmp",
        version = false,
        event = "InsertEnter",
        dependencies = {
            "hrsh7th/cmp-nvim-lsp",
            "hrsh7th/cmp-path",
            "hrsh7th/cmp-buffer",
        },
    },
    { "hrsh7th/cmp-nvim-lsp, lazy = true },
    { "hrsh7th/cmp-path", lazy = true },
    { "hrsh7th/cmp-buffer, lazy = true },
}
```

> Now that we have our plugins set, lets configure the plugin. The `nvim-cmp`

Reviewer: plug-ins?

            },
            -- Make 'enter' key select the completion
            mapping = cmp.mapping.preset.insert({
                ["<CR>"] = cmp.mapping.confirm({ select = true })
            }),
        }
        cmp.setup(opts)
    end,
},
```

Using the `tab` key to select completions is a fairly popular option, so we'll
go ahead and set that up now.

```lua
mapping = cmp.mapping.preset.insert({
    ["<CR>"] = cmp.mapping.confirm({ select = true }),
    ["<tab>"] = cmp.mapping(function(original)
        if cmp.visible() then
            cmp.select_next_item() -- run completion selection if completing
        else
            original()      -- run the original behavior if not completing
        end
    end, {"i", "s"}),
    ["<S-tab>"] = cmp.mapping(function(original)
        if cmp.visual() then
            cmp.select_prev_item()
        else
            original()
        end
    end, {"i", "s"}),
}),
```

Pressing `tab` while the completion menu is visible will select the next
completion and `shift` + `tab` will select the previous item. The tab behavior
falls back on whatever pre-defined behavior was there originally if the menu
isn't visible.

## Snippets

> https://github.com/L3MON4D3/LuaSnip

Reviewer: Consider adding a lead-in sentence to provide context and embed the GH link?

Author: While I'm not married to it, I was trying to make it very clear where the plugin is coming from. Each section starts with a link to the plugin doing the heavy-lifting so that it's clear where to look for docs, or just poke around in the plugin itself.

Reviewer: Ack. I think you can simply say that to set expectations and make it clear to users. Something like, "Each of these sections shows the link to the plugin to help you find the documentation or explore the plugin itself."

Reviewer 2: I agree, the link here seems like a typo

                        original()      -- run the original behavior if not completing
                    end
                end, {"i", "s"}),
                ["<S-tab>"] = cmp.mapping(function(original)
                    if cmp.visual() then
                        cmp.select_prev_item()
                    elseif luasnip.expand_or_jumpable() then
                        luasnip.jump(-1)
                    else
                        original()
                    end
                end, {"i", "s"}),
            }),
            snippets = {
                expand = function(args)
                    luasnip.lsp_expand(args)
                end,
            },
        }
        cmp.setup(opts)
    end,
},
```

Now our tab key is thoroughly overloaded in super-tab fashion; if the
completion window is open, pressing tab selects the next item in the list. If
you press tab over a snippet, the snippet will expand and continuing to press
tab results in jumping to the next insertion point. Finally, if you're neither
auto-completing nor expanding a snippet, it will behave like a normal `tab` key.

Now we need to write up some snippets. LuaSnip supports several snippet formats,

> including the popular TextMate format, VSCode format, its own Lua-based API, and

Reviewer: Consider writing VSCode as Visual Studio Code (VScode)

snippets coming from an LSP server.

Here are some snippets that I've found to be useful:

```snipmate
snippet pub "public access control"
  public $0

snippet priv "private access control"
  private $0

snippet if "if statement"
  if $1 {
    $2
  }$0

snippet ifl "if let"
  if let $1 = ${2:$1} {
    $3
  }$0

snippet ifcl "if case let"
  if case let $1 = ${2:$1} {
    $3
  }$0

snippet func "function declaration"
  func $1($2) $3{
    $0
  }

snippet funca "async function declaration"
  func $1($2) async $3{
    $0
  }

snippet guard
  guard $1 else {
    $2
  }$0

snippet guardl
  guard let $1 else {
    $2
  }$0

snippet main
  @main public struct ${1:App} {
    public static func main() {
      $2
    }
  }$0
```

Another popular snippet plugin worth mentioning is UltiSnips which allows you to
use inline Python while defining the snippet, greatly improving what kinds of
snippets you can write.

## Auto Commands

> Sometimes you'll want to open swift interface files or SIL files. By default,

Reviewer 3: Either "Swift interface" or "`.swiftinterface`".

```suggestion
Sometimes you may want to open `.swiftinterface` files or SIL files. By default,
```

Also, SIL files are probably a compiler-engineer-only game?

you won't get much assistance. We'll use autocommands to configure your editing
experience now.

Create a new file for you autocommand configurations.

```lua
-- lua/config/autocmds.lua

local augroup = vim.api.nvim_create_augroup
local autocmd = vim.api.nvim_create_autocmd

-- Filetypes
augroup("SetFileTypes", { clear = true })
autocmd({'Bufread', 'Bufnewfile'}, {
    group = 'SetFileTypes',
    pattern = "*.swiftinterface",
    command = 'setlocal ft=swift tw = 0',
})

autocmd({'Bufread', 'Bufnewfile'}, {
    group = 'SetFileTypes',
    pattern = "*.sil",
    command = 'setlocal ft=sil',
})
```

Then include that in your `init.lua`.

```
-- init.lua
...
require('config.autocmds')
```

> Restarting vim, you will have syntax highlighting for swift interface and SIL

Reviewer 3: ```suggestion
Restarting vim, you will have syntax highlighting for `.swiftinterface` and SIL
```

files.

# Conclusion

Swift development with Neovim is a solid experience once you have things

> configured correctly. This should give you foundation for building your

Reviewer 3: ```suggestion
configured correctly. This should give you a foundation for building your
```

development experience.

> TODO: Finish this

Reviewer 3: Finish this?

author: [etcwilde]
---

> Neovim is a modern reimplementation of the *vi text editor*, with a lot of new functionality piled

Author: Tbh, this should probably read vim instead. 🤔

Reviewer: @etcwilde - Spelled out "vim" on first use. Edited introduction to align tense (present), changed from first person to active voice, and reformatted for easier scannability per style guidelines.
